compiler/rustc_mir_transform/src/coroutine/relocate_upvars.rs

```diff
@@ -0,0 +1,451 @@
+//! MIR rewrite pass to promote upvars into native locals in the coroutine body
+//!
+//! # Summary
+//! This pass performs the following transformations.
+//! 1. It generates a fresh batch of locals for each captured upvars.
+//!
+//! For each upvar, whether used or not, a fresh local is created with the same type.
+//!
+//! 2. It replaces the places pointing into those upvars with places pointing into those locals instead
+//!
+//! Each place that starts with access into the coroutine structure `_1` is replaced with the fresh local as
+//! the base. For instance, `(_1.4 as Some).0` is rewritten into `(_34 as Some).0` when `_34` is the fresh local
+//! corresponding to the captured upvar stored in `_1.4`.
+//!
+//! 3. It assembles an prologue to replace the current entry block.
+//!
+//! This prologue block transfers every captured upvar into its corresponding fresh local, *via scratch locals*.
+//! The upvars are first completely moved into the scratch locals in batch, and then moved into the destination
+//! locals in batch.
```

Comment:
I do not understand what these "scratch" locals are doing or why they are needed. Could you add an example?